pkg/inject/context.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inject
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Context carries dependencies across component boundaries.
+type Context interface {
+	// PopulateThis populates the given pointer with an instance of its type,
+	// if available.  The target *must* be a pointer.  Key is an optional key
+	// to differentiate different values of the same type.  If it's the same
+	// as the type name, pass the empty string.
+	PopulateThis(key string, target interface{}) bool
+}
+
+// Nothing returns a new Context that doesn't provide any dependencies.
+func Nothing() Context {
+	return emptyContext{}
+}
+
+// emptyContext doesn't provide anything
+type emptyContext struct{}
+
+func (c emptyContext) PopulateThis(_ string, _ interface{}) bool { return false }
+
+// depContext is a Context that provides a dependency of a given type.
+// It'll call out to next if it can provide the type asked for.
+type depContext struct {
+	typ         reflect.Type
+	value       reflect.Value
+	key         string
+	keyOptional bool
+	next        Context
```

Comment:
wondering if having `depContext` as a container of multiple values can simplify things.